test/files/pos/InferOverloadedPartialFunction.scala

```diff
@@ -0,0 +1,41 @@
+class MySeq[T] {
+  def map1[U](f: T => U): MySeq[U] = new MySeq[U]
+  def map2[U](f: T => U): MySeq[U] = new MySeq[U]
+}
+
+class MyMap[A, B] extends MySeq[(A, B)] {
+  def map1[C](f: (A, B) => C): MySeq[C] = new MySeq[C]
+  def map1[C, D](f: (A, B) => (C, D)): MyMap[C, D] = new MyMap[C, D]
+  def map1[C, D](f: ((A, B)) => (C, D)): MyMap[C, D] = new MyMap[C, D]
+
+  def foo(f: Function2[Int, Int, Int]): Unit = ()
+  def foo[R](pf: PartialFunction[(A, B), R]): MySeq[R] = new MySeq[R]
+}
+
+object Test {
+  val m = new MyMap[Int, String]
+
+  // This one already worked because it is not overloaded:
+  m.map2 { case (k, v) => k - 1 }
+
+  // These already worked because preSelectOverloaded eliminated the non-applicable overload:
+  m.map1(t => t._1)
+  m.map1((kInFunction, vInFunction) => kInFunction - 1)
+  val r1 = m.map1(t => (t._1, 42.0))
+  val r1t: MyMap[Int, Double] = r1
+
+  // These worked because the argument types are known for overload resolution:
+  m.map1({ case (k, v) => k - 1 }: PartialFunction[(Int, String), Int])
+  m.map2({ case (k, v) => k - 1 }: PartialFunction[(Int, String), Int])
+
+  // These ones did not work before:
+  m.map1 { case (k, v) => k }
+  val r = m.map1 { case (k, v) => (k, k*10) }
+  val rt: MyMap[Int, Int] = r
+  m.foo { case (k, v) => k - 1 }
```

Comment:
You'd have to use `(m: MySeq[(Int, String)]).map1 { ... }`. In the case of `TreeSet`, 4 would take an additional implicit `Ordering`. Since implicits are not taken into account for overload resolution, you can't select an overload based on the presence or absence of an implicit:

```
Welcome to Scala 2.12.1 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_102).
Type in expressions for evaluation. Or try :help.

scala> class MySet[T] { def map[R](f: T => R): MySet[R] = new MySet[R] }
defined class MySet

scala> class MyTreeSet[T] extends MySet[T] { def map[R : Ordering](f: T => R): MyTreeSet[R] = new MyTreeSet[R] }
defined class MyTreeSet

scala> (new MyTreeSet[Int]).map(x => x)
res0: MyTreeSet[Int] = MyTreeSet@46e69805

scala> (new MyTreeSet[AnyRef]).map(x => x)
<console>:13: error: No implicit Ordering defined for AnyRef.
       (new MyTreeSet[AnyRef]).map(x => x)
                                  ^
```

If we want to support this we need a more elaborate scheme where the implicit parameter determines  the return type (so it couldn't be a simple `Ordering`), in other words: `CanBuildFrom`. I assume that's what @Ichoran had in mind when he suggested that CBF would still be useful for some operations that need additional flexibility, even if the base methods don't use it.